### PR TITLE
MM-22845: Added support for permalink previews.

### PR DESCRIPTION
**Pull request by mkraft**
#### Summary

* Adds metadata about the referenced post when a post is created or updated via the API or App where the first link in the message is a permalink.
* Caches the data into the post metadata cache.
* Only embeds the data if the current user has permissions to view the post.
* Adds a prop in the post record which saves the ID of the referenced post.
* Adds an instance-wide config setting to disable the embedding of permalink metadata.
* Fixed a few linting warnings unrelated to the ticket.

#### Ticket Link

https://mattermost.atlassian.net/browse/MM-22845

#### Related PRs

* https://github.com/mattermost/enterprise/pull/1005
* https://github.com/mattermost/mattermost-webapp/pull/8256

#### Release Note

```release-note
Embedded information about referenced post for first links that are permalinks.
```

### CODE REVIEW

**Comment by srkgupta** [comment]
Adding myself for review since some CodeQL alerts were generated on this PR.

File: model/client4.go
Comment (hanzei) [inline model/client4.go:2962]: Typo

**Comment by mkraft** [comment]
> Adding myself for review since some CodeQL alerts were generated on this PR. Will be reviewing the PR once the changes are completed.

The indentation of that LOC changed and it seems to have triggered an alert. What's the fix?

File: app/post_metadata.go
Comment (hmhealey) [inline app/post_metadata.go:483]: 0/5, but perhaps we could make a store method to get a PreviewPost in a single DB query

File: app/post_metadata.go
Comment (hmhealey) [inline app/post_metadata.go:456]: Since we're not persisting any information to the DB, should we maybe just skip this step? This branch also duplicates most of the code from below.

File: app/post_metadata.go
Comment (hmhealey) [inline app/post_metadata.go:181]: Should permalink previews also be behind a feature flag?

**Review by hmhealey — CHANGES_REQUESTED** [review]

**Comment by srkgupta** [comment]
Checked the CodeQL alert and found that it was a False positive and have dismissed the same. Removing myself from list of reviewers.

File: api4/post.go
Comment (mkraft) [reply api4/post.go:129]: Good question, there are at least 3 user ids in scope:
* `c.AppContext.Session().UserId`
* `ephRequest.Post.UserId`
* `ephRequest.UserID`
I think the session id is the correct one to use because it's not given as a user input; we don't want users to be able to provide a user ID and then have access to posts that they wouldn't normally be able to see.

File: app/post_metadata.go
Comment (esethna) [reply app/post_metadata.go:181]: @hmhealey yes good call. cc// @mkraft

File: app/post_metadata.go
Comment (mkraft) [reply app/post_metadata.go:225]: Good question. I copied this from the main place that checks this permissions but I haven't questioned whether just the first part is sufficient.

File: app/post_metadata.go
Comment (mkraft) [reply app/post_metadata.go:181]: Thanks for bringing this up. I was on the fence about adding a feature flag but I've added it because it's low effort and gives us the choice to disable or phase the rollout.

File: app/post.go
Comment (mkraft) [reply app/post.go]: You're right. Thanks for flagging this. I've made a change that customizes the WS message based on authz to the post in the case that the post is a permalink preview post.

File: app/post_metadata.go
Comment (mkraft) [reply app/post_metadata.go]: `PERMISSION_READ_PUBLIC_CHANNEL` is a team-scoped permission, so it has to be checked on the team, whereas `PERMISSION_READ_CHANNEL` is channel-scoped. Long story is this is how it needs to be done for now and how it's done elsewhere. However, I did add a helper method to make it a bit less wordy.

File: app/post_metadata.go
Comment (mkraft) [reply app/post_metadata.go:479]: That's a great idea that I'd like to defer to a future improvement for the sake of getting this out an iterating on it. I've added [a Jira story](https://mattermost.atlassian.net/browse/MM-37081) that links to this comment.

File: app/post_metadata.go
Comment (mkraft) [reply app/post_metadata.go:417]: We could maybe support relative links, but that may be an edge case for our community/community-daily case. Not sure how common that is. Where are you 0-5 that we should support relative links out of the gate?

What's the sneaky regex thing you're talking about? I didn't follow.

File: app/post_metadata.go
Comment (mkraft) [reply app/post_metadata.go:452]: This populates the permalink that eventually is returned as part of the JSON payload, I'm not sure how this could be skipped. Maybe I'm not seeing something?

**Comment by esethna** [comment]
Nice! @mkraft is there a way to test this front-end and backend together?

**Comment by mkraft** [comment]
> Nice! @mkraft is there a way to test this front-end and backend together?

Yes, @BenCookie95's all over it for you. We'll DM you details.

File: app/post_metadata.go
Comment (hmhealey) [reply app/post_metadata.go:417]: I'm 0/5 on whether or not we ship this without support for community, but relative link support is 100% necessary to have this work on community since it uses the relative links to work across servers in the cluster. But that also needs https://mattermost.atlassian.net/browse/MM-29224 anyway because you need to use a Markdown link to do a relative link.

As for the sneaky bit, I was trying to figure out if there was any way to trick the regex and have it return true for something that wasn't a valid permalink, but I couldn't come up with any way to actually abuse that, so it's probably not an issue.

File: api4/post.go
Comment (hmhealey) [reply api4/post.go:129]: Isn't `ephRequest.UserID` the one who would actually be receiving the post though? If I understand the code correctly, the first two IDs are the one making the ephemeral post and the third is the one receiving it, so I'd expect it to be done based on the recipient's user ID.

File: app/post_metadata.go
Comment (hmhealey) [reply app/post_metadata.go:452]: Maybe I'm misreading this, but the logic looks somewhat like
```
if the metadata is in the database:
    if the url is a permalink:
        load the post and return it

load the post and return it
```
so the information stored in the database isn't being used for permalinks because we're not actually retrieving anything about the preview post with `getLinkMetadataFromDatabase`. That's what's confusing me a bit about this code. I don't think it'll cause any problems, but the logic seems weird to me.

File: app/notification.go
Comment (hmhealey) [inline app/notification.go:472]: ~~Is this okay or should we be paging through this properly?~~

Never mind. We do this in a few other places. It looks kind of odd, but I guess we'll never have a channel with 10 million users. If we do end up with that, we definitely have bigger concerns than this breaking though.

File: app/notification.go
Comment (hmhealey) [inline app/notification.go:488]: Since these are mutating the message, do we need to copy the message in each of these loops? I'm thinking both because future iterations of the loop might want the non-cleared metadata, but also because we would potentially be getting some racy behaviour here assuming `a.Publish` does something asynchronous.

File: app/notification.go
Comment (hmhealey) [inline app/notification.go:487]: Not for this PR, but I feel like it would be really nice if this had a `UserIds` so we'd only have to potentially publish two messages here, one for users with permission and one for those without.

File: app/post_metadata.go
Comment (hmhealey) [inline app/post_metadata.go:221]: Why pass the empty string for new posts and updates? I think it might be safer to keep passing that since I think it gives us an extra layer of security by preventing us from potentially accidentally leaking some data if a user without permission to see a post is posting a permalink to it.

I don't expect us to hopefully leak stuff here, but I'm trying to make the default case that we don't send anything just in case something gets broken in the future.

**Review by hmhealey — CHANGES_REQUESTED** [review]

File: api4/post.go
Comment (mkraft) [reply api4/post.go:129]: Maybe... I'm a bit confused by the difference between `PostEphemeral.UserID` and `PostEphemeral.Post.UserId` (which is set to `c.AppContext.Session().UserId`).

File: app/post_metadata.go
Comment (mkraft) [reply app/post_metadata.go:417]: > relative link support is 100% necessary to have this work on community

Yes, totally agree. There was some talk about disabling that plugin, but I'll take a look at what's involved in supporting it out-of-the-gate and I'll either add it to this PR and we can review the changes or I'll add it as a separate ticket for afterwards.

File: app/notification.go
Comment (mkraft) [reply app/notification.go]: Yeah, only has a small performance for some peace of mind.

File: app/post_metadata.go
Comment (mkraft) [reply app/post_metadata.go:221]: Good question. There are absolutely other ways to achieve the same thing. It stems from the edge case that you might post a link to a post that you yourself can't access but others in the channel can.

This method puts the referenced post in the metatdata and *then* `nil`s it if for those who don't have authz. Alternatively, we could *not* put the referenced post in the metadata and then *add* it for those who have authz.

I totally see your point, you're concerned that someone might come along and use the app method with a blank `askingUserID` and expect to have the results authorized to the current user?

**Review by BenCookie95 — CHANGES_REQUESTED** [review]

File: api4/post.go
Comment (mkraft) [reply api4/post.go:403]: Perfect, I'll remove all of this code then.

File: app/post_metadata.go
Comment (mkraft) [reply app/post_metadata.go:429]: Recapping what we talked about offline: supporting the relative links used by community servers depends upon first [supporting markdown links](https://mattermost.atlassian.net/browse/MM-29224). When that's complete we can make a quick change to the regex so that relative permalinks embed the referenced post info.

File: api4/post.go
Comment (mkraft) [reply api4/post.go]: Recapping what we discussed: it looks like we have the correct user ID here.

File: app/post_metadata.go
Comment (mkraft) [reply app/post_metadata.go]: I've completely removed the knowledge of the "asking user" from this method. It now embeds the referenced posts' metadata (when appropriate) regardless of authz rules. A later "sanitization" method can optionally be invoked to invoke the authz rules for the given user.

**Review by michaelgamble — DISMISSED** [review]
Three minor UX issues:
1. post__preview-footer >> has double opacity on it, should just be the one (56%)
2. showMoreButton >> should be font-size 12 instead of 14
3. On hover of the post preview >> the dropshadow is supposed to change from "Elevation 1" to "Elevation 2" shadow style
 
To clarify point 3, only when hovering the preview card itself, not the whole post row.. so similar to how hovering over the preview link causes it to underline.. hovering over the card will cause a slight shift in the dropshadow.

/* Elevation 1 */
box-shadow: 0px 2px 3px rgba(0, 0, 0, 0.08);

/* Elevation 2 */
box-shadow: 0px 4px 6px rgba(0, 0, 0, 0.12);

**Comment by esethna** [comment]
@BenCookie95 I think these were meant to be posted on the webapp PR ^

File: app/post.go
Comment (mkraft) [reply app/post.go]: You're absolutely right, that's a good point. I did some refactoring to remove the duplication. Take a look and let me know your thoughts.

**Comment by stylianosrigas** [comment]
Removing the test server as it is failing to be created with the following error: Treatments: you passed PermalinkPreviews that does not exist in this environment, please double check what Splits exist in the web console

File: model/post_list.go
Comment (hmhealey) [inline model/post_list.go:30]: I think we also need to copy Order and Posts here. I'm pretty sure they're stored as something like pointers that aren't copied.

**Review by hmhealey — CHANGES_REQUESTED** [review]

File: app/post.go
Comment (mkraft) [reply app/post.go:722]: I think you're right that the WS message for the post shouldn't be interrupted simply because the referenced post (or channel) were deleted. I've update `publishWebsocketEventForPermalinkPost` to warn but not return the error if the referenced post or channel is 404.

**Review by hmhealey — APPROVED** [review]

**Review by BenCookie95 — APPROVED** [review]
LGTM, spent some time testing today!

**Comment by esethna** [comment]
@mkraft just checking, who are the required reviews for this PR? Can you help remove anyone whose review wouldn't block this from going in? I'd like to aim to merge this for v6.0 deadline

@jgilliam17 for QA review ^

cc// @amyblais

**Comment by mkraft** [comment]
> @mkraft just checking, who are the required reviews for this PR? Can you help remove anyone whose review wouldn't block this from going in? I'd like to aim to merge this for v6.0 deadline if possible (Aug 16)
> 
> @jgilliam17 for QA review ^
> 
> cc// @amyblais

I think @hanzei had just identified a typo but wasn't planning to be a reviewer and I think @michaelgamble had meant to be on the webapp pr.

**Review by esethna — APPROVED** [review]
Approving the server PR, webapp I think might need some pending changes as per: https://github.com/mattermost/mattermost-webapp/pull/8403#issuecomment-888482579

**Review by furqanmlk — COMMENTED** [review]
Thanks @BenCookie95 
Test cases passed except few, I mentioned below with red icon

<meta charset="utf-8"><b style="font-weight:normal;" id="docs-internal-guid-fe8da6cd-7fff-ef59-c9b5-5b7aa1a62562"><div dir="ltr" style="margin-left:-40.5pt;" align="left">

Test Case
--
Permalink Preview - Short Text ✅  
Permalink Preview - Short Text - Hover Post  ✅

</div></b>
<meta charset="utf-8"><b style="font-weight:normal;" id="docs-internal-guid-f919ef58-7fff-e413-754a-a0336a325fa2"><div dir="ltr" style="margin-left:-40.5pt;" align="left">

Permalink Preview - Short Text - Hover Preview  ✅
Permalink Preview - Truncated ✅

</div></b>

<meta charset="utf-8"><b style="font-weight:normal;" id="docs-internal-guid-24a8e543-7fff-507b-0f8a-dc997c0ed91c"><div dir="ltr" style="margin-left:-40.5pt;" align="left">

Nested Permalink ✅ 
Permalink Preview - image🔴  **( FAILED )** --> **No Images in Preview**
Permalink Preview - Multiple Attachments🔴  **( FAILED )** **No Files in Preview**

</div></b>
<meta charset="utf-8"><b style="font-weight:normal;" id="docs-internal-guid-2e34ecb3-7fff-9402-8d7d-c23599403f80"><div dir="ltr" style="margin-left:-40.5pt;" align="left">

Permalink Preview - DM’s  ✅
Permalink Preview - Private channels ✅
Permalink Preview - Guest 🔴  **( FAILED )** --> **Guest can see and open Permalink Preview**

</div></b>
<meta charset="utf-8"><b style="font-weight:normal;" id="docs-internal-guid-b4854db6-7fff-cbfd-68d2-a4711939fa9a"><div dir="ltr" style="margin-left:-40.5pt;" align="left">

Permalink Preview - Public into Private Channel  ✅
Permalink Preview - Edited ✅
Permalink Preview -Deleted ✅

</div></b>

**Comment by mkraft** [comment]
> Permalink Preview - image🔴 **( FAILED )** --> **No Images in Preview**

That is coming in a future iteration.

> Permalink Preview - Multiple Attachments🔴 **( FAILED )** **No Files in Preview**

Coming in a future iteration.

> Permalink Preview - Guest 🔴 **( FAILED )** --> **Guest can see and open Permalink Preview**

My understanding is that guests *should* see permalink previews if they have the appropriate authorization to the referenced post.

cc @esethna @BenCookie95

**Comment by esethna** [comment]
> My understanding is that guests should see permalink previews if they have the appropriate authorization to the referenced post.

That's correct @mkraft @furqanmlk

**Comment by esethna** [comment]
@furqanmlk to clarify your testing results: For the DM and private channel case, can you confirm that the observed behavior was that previews were shown to members that have access to the DM/GM/private channel?

**Comment by furqanmlk** [comment]
> @furqanmlk to clarify your testing results: For the DM and private channel case, can you confirm that the observed behavior was that previews are shown to members that have access to the DM/GM/private channel?

Yes @esethna , that's correct, `Permalink Preview is shown to the member who have access to the DM/GM/Private Channels.`
[Test plan](https://docs.google.com/document/d/1HvfQKgABCKO9sXtGxcvykI4_ZZVT7IYFlvwTAvQ3d0M/edit#)

**Comment by esethna** [comment]
@furqanmlk thanks, does this pass QA in that case? Please add your approval on all the PRs if so, thanks!

**Comment by furqanmlk** [comment]
Hi @esethna 
All tests passed except one that I posted in other PR

https://github.com/mattermost/enterprise/pull/1005

Martin is looking at this. I will approve as soon as I hear from him.

**Comment by mkraft** [comment]
> Hi @esethna
> All tests passed except one that I posted in other PR
> 
> [mattermost/enterprise#1005](https://github.com/mattermost/enterprise/pull/1005)
> 
> Martin is looking at this. I will approve as soon as I hear from him.

We're going to address the 2nd post created as a separate bug.

**Review by furqanmlk — APPROVED** [review]

**Pull request merged**
